core/src/main/scala/shapeless/package.scala

```diff
@@ -59,6 +59,60 @@ package object shapeless {
   type ∃[P[_]] = P[T] forSome { type T }
   type ∀[P[_]] = ¬[∃[({ type λ[X] = ¬[P[X]]})#λ]]
 
+  //Type logic-gates
+  trait ||[A, B] extends Serializable {
+    def a: Option[A]
+    def b: Option[B]
+  }
+
+  object || {
+    implicit def aOrTpe[A, B](implicit eva: A): A || B = new ||[A, B] {
+      val a = Some(eva)
+      val b = None
+    }
+    implicit def bOrTpe[A, B](implicit evb: B): A || B = new ||[A, B] {
+      val a = None
+      val b = Some(evb)
+    }
+  }
+
+  implicit def abOrTpe[A, B](implicit eva: A, evb: B): A || B = new ||[A, B] {
```

Comment:
I think this works because the package object is counted as the local scope and is a separate scope to the object so it stops searching after the first implicit in the `&&` case. I can check in a separate project though with a local build to make sure though.